### ISSUE TITLE
[Feature] Make Filter structure in Search package more extendible

### ISSUE THREAD

**Issue by rohithbalaji123** [short-d/short#879]
### What is the problem?
There is an inherent coupling between Resource and Order structures in [Filter](https://github.com/short-d/short/blob/master/backend/app/usecase/search/filter.go) and it is quite difficult to extend when we try to add more resources or order by classes. That is., there is a requirement for every Resource to support each Order bys which will not be the case at all times.

#### When trying to add a new `Resource`:
- We will need to add a method to the [Order](https://github.com/short-d/short/blob/master/backend/app/usecase/search/order/order.go#L13) interface with the name `ArrangeResource`. That would mean that we need to implement those methods in all existing structures which implements `Order`(say [CreatedTime](https://github.com/short-d/short/blob/master/backend/app/usecase/search/order/created_time.go#L13) ). But, the newly added Resource may not even have any field related to created time.

#### When trying to add a new `Order`:
- When we have a requirement to add a new Order By(say a new column which tracks number of times the short link is visited), we cannot add this [Order By](https://github.com/short-d/short/blob/master/backend/app/usecase/search/order/order.go#L6) as that would mean, we will have to implement all the methods in the `Order` interface (currently has `ArrangeShortLinks`, `ArrangeUsers`). But, this new Order By is related only to ShortLinks.

### Your solution
Will have to think through to come up with a more generic solution.

### Alternatives considered
For resources not supporting a particular order by, we can throw `NotApplicable` error when the package's client tries to order by the unsupported order by. This will also include validating in [NewFilter](https://github.com/short-d/short/blob/master/backend/app/usecase/search/filter.go#L25) that the resource and order arrays are compatible with each other.

**Comment by byliuyang** [comment]
@rohithbalaji123 
1) I don't think all resources has to support the same order. When can do unchanged order for those not supported.
2) We don't have generics support in Go 1. Waiting for Go 2. It's already in experimentation stage

**Comment by rohithbalaji123** [comment]
@byliuyang Yes, we can do unchanged order for unsupported resource. My concern is, what should `ArrangeNewResource` method's implementation be in [CreatedTime](https://github.com/short-d/short/blob/master/backend/app/usecase/search/order/created_time.go#L13) structure when `NewResource` doesn't have a column called `createdAt`?

**Comment by byliuyang** [comment]
@rohithbalaji123 It doesn’t do anything? Is this avoidable?

**Comment by rohithbalaji123** [comment]
@byliuyang , I am not sure how to avoid it. When we add more and more resources and Order By implementations, there will be a lot combinations where the method implementations will be empty.

**Comment by arberiii** [comment]
We can also reverse things here, so each resource implements each order, and in case that an implementation is not possible we keep the slice unchanged. So we can have a resource type []entity.ShortLink which implements all the order types
